Add log upload support for build target failed using teatime.py script

erman-gurses opened the pull request:
This PR integrates logging mechanism with` teatime.py` script. 
Also handles the TODO below:
 `# TODO: Fork indexer.py locally to avoid relying on an external GitHub source at runtime.`

erman-gurses commented:
Last Tests:
Linux: https://github.com/ROCm/TheRock/actions/runs/15155118087
Windows: https://github.com/ROCm/TheRock/actions/runs/15155125988

marbre commented:
Is this ready to review or do you just want initial feedback as this is still marked as a draft. Regarding the TODO in the PR description: I would rather replace `indexer.py` with an AWS Lambda and use boto3.

erman-gurses commented:
> Is this ready to review or do you just want initial feedback as this is still marked as a draft. Regarding the TODO in the PR description: I would rather replace `indexer.py` with an AWS Lambda and use boto3.

It is ready to review - I can test with boto3.

erman-gurses commented:
> > Is this ready to review or do you just want initial feedback as this is still marked as a draft. Regarding the TODO in the PR description: I would rather replace `indexer.py` with an AWS Lambda and use boto3.
> 
> It is ready to review - I just make it ready PR from WIP - I can test with boto3.

@marbre @ScottTodd boto3 gives error on Windows side:
```
Traceback (most recent call last):
  File "C:\home\runner\_work\TheRock\TheRock\build_tools\create_log_index.py", line 6, in <module>
    import boto3
ModuleNotFoundError: No module named 'boto3'
[WARN] create_log_index.py failed: Command '['C:\\home\\runner\\_work\\_tool\\Python\\3.12.10\\x64\\python3.exe', 'C:\\home\\runner\\_work\\TheRock\\TheRock\\build_tools\\create_log_index.py', '--build-dir', 'B:\\tmpbuild', '--amdgpu-family', 'gfx110X-dgpu']' returned non-zero exit status 1.
```

ScottTodd commented:
Is the workflow installing the package? It could be either explicitly installed with `pip install boto3` or we could include it in a requirements.txt file.

erman-gurses commented:
> Is the workflow installing the package? It could be either explicitly installed with `pip install boto3` or we could include it in a requirements.txt file.

Sounds good! I can install it on the workflow.

marbre commented:
> > Is this ready to review or do you just want initial feedback as this is still marked as a draft. Regarding the TODO in the PR description: I would rather replace `indexer.py` with an AWS Lambda and use boto3.
> 
> It is ready to review - I just make it ready PR from WIP - I can test with boto3.

Taking a quick look: Let's not solve too many problems at once! For now I would replace uploading files via AWS cli by using boto3. The AWS Lambda is a different thing and will replace `indexer.py` but that should be done in a separate PR. What I was saying is that the TODO note is probably not what we aim for but as said that's something for a follow up.

erman-gurses commented:
> > > Is this ready to review or do you just want initial feedback as this is still marked as a draft. Regarding the TODO in the PR description: I would rather replace `indexer.py` with an AWS Lambda and use boto3.
> > 
> > 
> > It is ready to review - I just make it ready PR from WIP - I can test with boto3.
> 
> Taking a quick look: Let's not solve too many problems at once! For now I would replace uploading files via AWS cli by using boto3. The AWS Lambda is a different thing and will replace `indexer.py` but that should be done in a separate PR. What I was saying is that the TODO note is probably not what we aim for but as said that's something for a follow up.

Sounds good to me!  I can raise AWS Lambda. So I am using the indexer as it was and just updating AWS cli by using boto3.

erman-gurses commented:
> > > > Is this ready to review or do you just want initial feedback as this is still marked as a draft. Regarding the TODO in the PR description: I would rather replace `indexer.py` with an AWS Lambda and use boto3.
> > > 
> > > 
> > > It is ready to review - I just make it ready PR from WIP - I can test with boto3.
> > 
> > 
> > Taking a quick look: Let's not solve too many problems at once! For now I would replace uploading files via AWS cli by using boto3. The AWS Lambda is a different thing and will replace `indexer.py` but that should be done in a separate PR. What I was saying is that the TODO note is probably not what we aim for but as said that's something for a follow up.
> 
> Sounds good to me! I can raise another PR for AWS Lambda. So I am using the indexer as it was and just updating AWS cli by using boto3.

The last tests:
Linux:        https://github.com/ROCm/TheRock/actions/runs/15177259483/job/42679694420
Windows: https://github.com/ROCm/TheRock/actions/runs/15177261545

marbre commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -47,26 +54,32 @@ def upload_logs_to_s3(s3_base_path: str, build_dir: Path):
+        log(f"[ERROR] Invalid s3_base_path: {s3_base_path}")
+        sys.exit(2)
+
+    bucket, *prefix_parts = s3_base_path[5:].split("/", 1)
```
I think this might be error prone depending on the endpoint. Maybe go closer with what I've done in `generate_release_index.py`:
https://github.com/ROCm/TheRock/blob/b354440c1e585d2aeca63209f2b9b4b5dac5e3a3/build_tools/packaging/python/generate_release_index.py#L30-L45

erman-gurses replied:
Updated the code.

marbre commented on build_tools/teatime.py:
```diff
@@ -103,6 +110,60 @@ def finish(self):
+        # Step 2: Upload logs to S3
+        try:
+            upload_script = repo_root / "build_tools" / "upload_logs_to_s3.py"
+            s3_path = f"s3://{self.s3_bucket}/{self.s3_subdir}"
```
I would rather pass the explicit arguments instead of fusing and splitting them later again. See other comment.

marbre commented on build_tools/teatime.py:
```diff
@@ -103,6 +110,60 @@ def finish(self):
+                ],
+                check=True,
+            )
+        except subprocess.CalledProcessError as e:
+            print(f"[WARN] Log upload failed: {e}", file=sys.stderr)
+        except Exception as e:
+            print(f"[WARN] Unexpected error during log upload: {e}", file=sys.stderr)
```
What are the assumption with regards to uploading logs? Is it purely optional or is it expected to work when the environment variables are set accordingly? The minimum is to document the behavior. Anyway, I would vote for the latter as it otherwise hard to catch any error and we use this script in the CI. We don't want to notice that logs are missing if we really need them. Another option would be to add a setting to control if the script should fail if the log upload did not succeeded.

erman-gurses replied:
We have the environment variables below -  if they are not set -  uploading will not work
```
      TEATIME_S3_UPLOAD: "1"
      TEATIME_S3_BUCKET: "therock-artifacts"
      TEATIME_S3_SUBDIR: "${{ github.run_id }}-windows/logs/${{ inputs.amdgpu_families }}"
```

marbre replied:
Yes, but this only gets you a warning and wont let the script failing or did I miss something?

erman-gurses replied:
I see your point - but I need to ask more questions to make it crystal.
Do we want to fail the script if a single or multiple log files cannot be uploaded?
My concern is don't we want the script continue uploading the rest of the log files even some of them missing.

erman-gurses replied:
That might be an option?
```
except subprocess.CalledProcessError as e:
    print(f"[WARN] Log upload failed: {e}", file=sys.stderr)
    if os.getenv("TEATIME_FAIL_ON_UPLOAD_ERROR") == "1":
        raise
```

erman-gurses replied:
Updated the code accordingly

erman-gurses commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -2,42 +2,49 @@
+            return
+
+    if already_exists:
+        return  # Skip uploading and printing success log
```
The purpose of it is to avoid abundant log messages when we upload logs. The log files are updated partially ever time and after some point the log messages dominate the screen . To avoid that I still upload the files (updated files) not showing them in the log messages The remaining code was below and now. The new files, first time uploaded, I show them in the log messages:

The remaining code is updated. The old code does not allow to update the files that are already exist in the bucket.
```
    if already_exists:
        # Upload anyway for the additional logs,
        # but do not repeat success log
        try:
            s3.upload_file(str(file_path), bucket, key, ExtraArgs=extra_args)
        except ClientError as e:
            log(f"[ERROR] Re-upload failed for {file_path}: {e}")
        return  # Suppress success log
```

marbre commented on build_tools/create_log_index.py:
```diff
@@ -21,12 +21,44 @@ def normalize_path(p: Path) -> str:
     return str(p).replace("\\", "/") if is_windows() else str(p)
 
 
+def download_indexer(build_dir: Path) -> Path:
```
I think this comment rather intended to say that we need to fork the script. If we cannot download in the workflow (due to network issues), we probably also cannot download while running the script. Checking if the script is available and erroring otherwise is good but I wouldn't do the download in the script.

erman-gurses replied:
That makes sense -  will update the code

erman-gurses replied:
Done

erman-gurses commented on build_tools/teatime.py:
```diff
@@ -103,6 +110,60 @@ def finish(self):
 
+        # Return if upload is not configured
+        if not (self.upload_to_s3 and self.s3_bucket and self.s3_subdir):
+            return
```
That is a great idea -  updated

erman-gurses commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -47,26 +54,32 @@ def upload_logs_to_s3(s3_base_path: str, build_dir: Path):
 
+    if not s3_base_path.startswith("s3://"):
+        log(f"[ERROR] Invalid s3_base_path: {s3_base_path}")
+        sys.exit(2)
```
New at boto3 -  removed the redundant check.

ScottTodd commented:
@erman-gurses if you want a code review, please batch your changes and comments more or make the PR smaller. I can't follow the stream of commits and multiple review threads.

If you are still working on this, please keep the changes local or convert to a draft PR.

erman-gurses commented:
> @erman-gurses if you want a code review, please batch your changes and comments more or make the PR smaller. I can't follow the stream of commits and multiple review threads.
> 
> If you are still working on this, please keep the changes local or convert to a draft PR.

@ScottTodd you are so right -  will be more careful on that (have to break my habit).  My changes are done for now to address Marius's comments.

ScottTodd commented:
(Where by "If you are still working on this," I mean "if this is not yet ready for review")

erman-gurses commented:
> (Where by "If you are still working on this," I mean "if this is not yet ready for review")

Tried to say I addressed all comments from Marius and would like to take additional comments.

erman-gurses commented:
Windows: https://github.com/ROCm/TheRock/actions/runs/15268948168
Linux: https://github.com/ROCm/TheRock/actions/runs/15268938443
If these tests are passed, it is ready to review.

ScottTodd commented on build_tools/create_log_index.py:
```diff
@@ -21,22 +21,57 @@ def normalize_path(p: Path) -> str:
+    sys.exit(2)
+
+
 def index_log_files(build_dir: Path, amdgpu_family: str):
     log_dir = build_dir / "logs"
```
Here the script should be given the path to the logs directory and shouldn't need to assume that logs are always in `${build_dir}/logs`

erman-gurses replied:
@ScottTodd, How can I find out the log path?

ScottTodd replied:
CMake always sets to `${THEROCK_BINARY_DIR}/logs`, so that is safe to assume: https://github.com/ROCm/TheRock/blob/46ba6a0b37ba7abd2f3cbe43cfe995f170bd15b3/cmake/therock_subproject.cmake#L118-L121

This script looks like it only cares about the logs directory though, so it should be passed only the path to the logs directory, not the build directory.

ScottTodd commented on build_tools/teatime.py:
```diff
@@ -81,6 +81,12 @@ def __init__(self, args: argparse.Namespace):
 
+        # S3 upload configuration
+        self.upload_to_s3 = os.getenv("TEATIME_S3_UPLOAD", "0") == "1"
+        self.s3_bucket = os.getenv("TEATIME_S3_BUCKET")
+        self.s3_subdir = os.getenv("TEATIME_S3_SUBDIR")
+        self.base_build_dir = os.getenv("BASE_BUILD_DIR")
```
Might as well check that if one of `s3_bucket` or `s3_subdir` is set, the other must also be set, rather than silently fail below in that case:

```python
        # Call coordinate_index_and_logs if upload is configured
        if self.upload_to_s3 and self.s3_bucket and self.s3_subdir:
            self.coordinate_index_and_logs()
```

Actually, looking at this more closely, we shouldn't need `base_build_dir` here, but we _do_ need `log_path`. The `--build-dir` argument in `create_log_index.py` should be changed to `--logs-dir`. Rather than pass around the _build_ directory. That should result in some good simplification across the scripts and yaml files.

ScottTodd commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -2,45 +2,63 @@
 
 
-def upload_logs_to_s3(s3_base_path: str, build_dir: Path):
+def upload_logs_to_s3(bucket: str, subdir: str, build_dir: Path):
     log_dir = build_dir / "logs"
```
Here also, take `--logs-dir` instead of `--build-dir`

erman-gurses replied:
Updated

ScottTodd commented on build_tools/teatime.py:
```diff
@@ -81,6 +81,12 @@ def __init__(self, args: argparse.Namespace):
         self.log_timestamps: bool = args.log_timestamps
 
+        # S3 upload configuration
+        self.upload_to_s3 = os.getenv("TEATIME_S3_UPLOAD", "0") == "1"
```
Document all new options in the comment at the top of this file.

ScottTodd commented on build_tools/teatime.py:
```diff
@@ -103,6 +110,66 @@ def finish(self):
                 b"[" + self.label + b" completed in " + run_pretty.encode() + b"]\n"
             )
 
+        # Call coordinate_index_and_logs if upload is configured
+        if self.upload_to_s3 and self.s3_bucket and self.s3_subdir:
+            self.coordinate_index_and_logs()
```
teatime.py is called once _per subproject_: https://github.com/ROCm/TheRock/blob/25516d0df8821844845670d59a0bf8e18af66390/cmake/therock_subproject.cmake#L97-L109

Is this index page generation and uploading safe to run multiple times? Would it make more sense to generate the index page at the start or end of the build? I think ideally there would be a dynamic index page running on the server that would be aware of which logs have already been uploaded, rather than need to regenerate a static index page 10-20 times.

erman-gurses commented on .github/workflows/build_linux_packages.yml:
```diff
@@ -91,9 +96,25 @@ jobs:
+      - name: Install python and other deps
         run: |
           pip install -r requirements.txt
+          pip install boto3
```
Sounds good - added

erman-gurses commented:
@ScottTodd, regarding this PR - any other comments that I can address?

erman-gurses commented:
> > @ScottTodd, regarding this PR - any other comments that I can address?
> 
> It's in my queue. Patience :)

Thanks @ScottTodd

ScottTodd commented on build_tools/teatime.py:
```diff
@@ -103,6 +144,79 @@ def finish(self):
                 b"[" + self.label + b" completed in " + run_pretty.encode() + b"]\n"
             )
 
+        # Call coordinate_index_and_logs if indexing is enabled.
+        if (
+            not self.skip_index
+            and self.upload_to_s3
+            and self.s3_bucket
+            and self.log_dir
+        ):
+            self.coordinate_index_and_logs()
+
+    def coordinate_index_and_logs(self):
+        # Determine paths
+        log_dir = self.log_dir
+        amdgpu_family = os.getenv("AMDGPU_FAMILIES")
```
I'm not totally convinced that this indexing code should be in teatime.py. Also, as mentioned before, the logging script should not need to know low level details about individual build options like `AMDGPU_FAMILIES`.

There are two scenarios to design for:

1. CI usage, where we run a controlled sequence of commands (install deps, configure credentials, load from cache, build, report statistics, upload results, trigger other jobs)
2. Developer usage, where a developer just runs a CMake build.

The goal with this sort of integration work is to make it easier for both scenarios to upload logs to the cloud. At the moment, teatime.py is only responsible for creating individual log files. The cloud logging _also_ has the wrinkle of having a generated static index page for the logs directory. What we really should have IMO is a server with dynamic index pages _or_ some server-side processing of the uploaded logs, instead of pushing that responsibility on the client here.

We could look into having a CMake POST_BUILD (https://cmake.org/cmake/help/latest/command/add_custom_command.html) command that generates and uploads the index page after building `therock-archives`... or we could add a new helper target like `therock-logs-upload`. Those would integrate into the build system itself at some "finalization" point so developers could get their cmake build command to automatically upload in a similar way to how Bazel/Blaze can upload to resultstore/sponge, but that would still be only working around the fact that these index pages are statically generated.

ScottTodd replied:
Chatting with @marbre . We can move the index generation to an AWS lambda so whenever logs are uploaded to the bucket the script will run server-side. There are other ways to solve that it in the cloud, but the general approach is that we shouldn't need to generate index pages explicitly on the client side at all.

erman-gurses replied:
Sounds good, so no index generation in the workflow.

ScottTodd commented on build_tools/teatime.py:
```diff
@@ -178,6 +292,9 @@ def main(cl_args: list[str]):
         default=False,
         help="Log timestamps along with log lines to the log file",
     )
+    p.add_argument(
+        "--skip-index", action="store_true", help="Skip indexing and uploading logs"
+    )
```
I don't see value in having this argument in teatime.py. It's a big mode switch to have `cmake/therock_subproject.cmake` run one command and then `.github/workflows/build_*_packages.yml` run another command. At that point just have the workflow files run the index generator and upload script directly.

If keeping it, flipping the condition from opt-out to opt-in would make more sense to me. That way, running the script on a single file would not generate an index page.

ScottTodd commented on build_tools/teatime.py:
```diff
@@ -25,6 +25,18 @@
+  indexing/upload scripts.
+* AMDGPU_FAMILIES: Required by `create_log_index.py` to organize logs.
 
 CI systems can set `TEATIME_LABEL_GH_GROUP=1` in the environment, which will
```
All environment variables should be explained in the same section

ScottTodd commented on build_tools/create_log_index.py:
```diff
@@ -21,22 +21,41 @@ def normalize_path(p: Path) -> str:
 
-def index_log_files(build_dir: Path, amdgpu_family: str):
-    log_dir = build_dir / "logs"
+def get_indexer_path() -> Path:
+    """
+    Resolve to the local third-party/indexer/indexer.py copy.
+    """
+    indexer_path = (
+        Path(__file__).resolve().parent.parent / "third-party/indexer/indexer.py"
+    )
+    if indexer_path.is_file():
+        log(f"[INFO] Using bundled indexer.py: {indexer_path}")
+        return indexer_path
+    else:
+        log(f"[ERROR] Bundled indexer.py not found at: {indexer_path}")
+        sys.exit(2)
```
The else clause should never be hit. I would drop the error handling here.

ScottTodd commented on .github/workflows/build_linux_packages.yml:
```diff
@@ -91,9 +97,18 @@ jobs:
           pip install -r requirements.txt
+          pip install -r requirements-dev.txt
+
+      - name: Configure AWS Credentials
+        if: always()
+        uses: aws-actions/configure-aws-credentials@ececac1a45f3b08a01d2dd070d28d111c5fe6722 # v4.1.0
```
@geomin12 we're getting a lot of these `if: always()` and `if: ${{ !cancelled() }}` conditions in these workflows. Let's see if we can restructure the flow a bit through migrating to more scripts such that the flow is easier to understand.

Here in particular, if fetching sources fails, we'd still configure AWS credentials and try to upload logs, but we wouldn't try to build. In that case I would prefer for the workflow to fail fast and not continue to try cleanup steps that will fail.

ScottTodd commented on build_tools/teatime.py:
```diff
@@ -103,6 +144,79 @@ def finish(self):
+        else:
+            print("[WARN] AMDGPU_FAMILIES not set; skipping log indexing")
+
+        # Step 2: S3 upload using --bucket and --subdir
+        try:
+            upload_script = (
+                Path(__file__).resolve().parent.parent
+                / "build_tools"
+                / "upload_logs_to_s3.py"
+            )
+            print(f"[TEATIME] Uploading logs to s3://{self.s3_bucket}/{self.s3_subdir}")
+            subprocess.run(
+                [
+                    sys.executable,
+                    str(upload_script),
+                    "--log-dir",
+                    str(log_dir),
+                    "--bucket",
+                    self.s3_bucket,
+                    "--subdir",
+                    self.s3_subdir,
+                ],
+                check=True,
+            )
+        except subprocess.CalledProcessError as e:
+            print(f"[WARN] Log upload failed: {e}", file=sys.stderr)
+            if os.getenv("TEATIME_FAIL_ON_UPLOAD_ERROR") == "1":
+                raise
+        except Exception as e:
+            print(f"[WARN] Unexpected error during log upload: {e}", file=sys.stderr)
+            if os.getenv("TEATIME_FAIL_ON_UPLOAD_ERROR") == "1":
+                raise
```
We don't need to `subprocess.run` Python scripts we control from other Python scripts we control. Could switch to a Python library model where we import the file and call a function directly.

ScottTodd reviewed (changes requested):
I'm still not sure about the architecture here. Probably want more feedback from Marius when he gets back too.

erman-gurses commented:
Closing this since outdated -  will be open if there is any remaining development is needed.
